src/cephadm/cephadmlib/node_proxy/baseredfishsystem.py

```diff
@@ -0,0 +1,152 @@
+import concurrent.futures
+from .basesystem import BaseSystem
+from .redfish_client import RedFishClient
+from threading import Thread, Lock
+from time import sleep
+from .util import Logger, retry
+from typing import Dict, Any, List
+
+
+class BaseRedfishSystem(BaseSystem):
+    def __init__(self, **kw: Any) -> None:
+        super().__init__(**kw)
+        self.common_endpoints: List[str] = kw.get('common_endpoints', ['/Systems/System.Embedded.1',
+                                                                       '/UpdateService'])
+        self.chassis_endpoint: str = kw.get('chassis_endpoint', '/Chassis/System.Embedded.1')
+        self.log = Logger(__name__)
+        self.host: str = kw['host']
+        self.port: int = kw['port']
+        self.username: str = kw['username']
+        self.password: str = kw['password']
+        # move the following line (class attribute?)
+        self.client: RedFishClient = RedFishClient(host=self.host, port=self.port, username=self.username, password=self.password)
+        self.log.logger.info(f'redfish system initialization, host: {self.host}, user: {self.username}')
+
+        self.run: bool = False
+        self.thread: Thread
+        self.data_ready: bool = False
+        self.previous_data: Dict = {}
+        self.lock: Lock = Lock()
+        self.data: Dict[str, Dict[str, Any]] = {}
+        self._system: Dict[str, Dict[str, Any]] = {}
+        self._sys: Dict[str, Any] = {}
+        self.start_client()
+
+    def start_client(self) -> None:
+        self.client.login()
+        self.start_update_loop()
+
+    def start_update_loop(self) -> None:
+        self.run = True
+        self.thread = Thread(target=self.update)
+        self.thread.start()
+
+    def stop_update_loop(self) -> None:
+        self.run = False
+        self.thread.join()
+
+    def update(self) -> None:
+        #  this loop can have:
+        #  - caching logic
+        while self.run:
+            self.log.logger.debug('waiting for a lock.')
+            self.lock.acquire()
+            self.log.logger.debug('lock acquired.')
+            try:
+                self._update_system()
+                self._update_sn()
+                update_funcs = [self._update_memory,
+                                self._update_power,
+                                self._update_fans,
+                                self._update_network,
+                                self._update_processors,
+                                self._update_storage,
+                                self._update_firmwares]
+
+                with concurrent.futures.ThreadPoolExecutor() as executor:
+                    executor.map(lambda f: f(), update_funcs)
+
+                self.data_ready = True
+                sleep(5)
```

Comment:
what's the reason for this sleep?

Reply:
to be honest: I have no idea, probably a left over from a debug or something like that.

Reply:
oh, actually it seems it has always been there if you look at the initial commit [here](https://github.com/ceph/ceph/pull/54742/commits/551a4cad9be49ca47a0b8f119dc27d116f9733d7#diff-979cf291b66a12be27f02a0b28138103fdc54a12a53f76a4c681616cabc9e21eR91)
(probably to have a small break before each iteration in order to not overload the redfish api or something like that?)